### These are the most important steps to follow while dealing with tools

In [2]:
from langchain_core.tools import tool
from langchain_ollama import ChatOllama
import requests
from langchain_core.messages import HumanMessage

### Tool Creation

In [3]:
@tool
def multiply(a: int, b: int) ->int :
    """This function takes two integer arguments and will return product as an integer"""
    return a * b

In [4]:
llm = ChatOllama(
    base_url="http//:localhost:11434",
    model = "llama3.2",
    temperature=0.3
)


### Tool Binding

In [5]:
llm_with_tools = llm.bind_tools([multiply])

### Tool Calling

In [6]:
llm_with_tools.invoke("What is 2 * 4")

AIMessage(content='', additional_kwargs={}, response_metadata={'model': 'llama3.2', 'created_at': '2026-04-24T09:19:50.0534355Z', 'done': True, 'done_reason': 'stop', 'total_duration': 11008163800, 'load_duration': 9471659700, 'prompt_eval_count': 162, 'prompt_eval_duration': 932751400, 'eval_count': 22, 'eval_duration': 526300300, 'logprobs': None, 'model_name': 'llama3.2', 'model_provider': 'ollama'}, id='lc_run--019dbec9-a3e5-7b10-8e7c-6b85aa777823-0', tool_calls=[{'name': 'multiply', 'args': {'a': 2, 'b': 4}, 'id': 'ba8061c4-8bc6-4485-b537-f2e58762d75d', 'type': 'tool_call'}], invalid_tool_calls=[], usage_metadata={'input_tokens': 162, 'output_tokens': 22, 'total_tokens': 184})

### When we invoke our llm, the llm will not call a tool, instead it gives us an object wich contains the required data for tool as we can see in ex below:  
tool_calls=[{'name': 'multiply', 'args': {'a': 2, 'b': 4}, 'id': 'a9e2f08c-e437-4f05-8da6-1b1c59114bb1', 'type': 'tool_call'}], invalid_tool_calls=[], usage_metadata={'input_tokens': 162, 'output_tokens': 22, 'total_tokens': 184})

In [7]:
llm_with_tools.invoke("what is the product of 1000 and 3").tool_calls

[{'name': 'multiply',
  'args': {'a': 1000, 'b': 3},
  'id': '00d304f8-8f42-49e4-845e-6d9823670df3',
  'type': 'tool_call'}]

### Tool Execution

In [8]:
result = llm_with_tools.invoke("what is the product of 1000 and 3").tool_calls

In [18]:
print(multiply.invoke(result[0]["args"]))

3000


### Professional way of tool calling with history

In [20]:
query = HumanMessage("Can you multiply 5 and 25")

In [21]:
messages = [query]

In [24]:
result = llm_with_tools.invoke(messages)
result

AIMessage(content='', additional_kwargs={}, response_metadata={'model': 'llama3.2', 'created_at': '2026-04-24T09:36:58.8260125Z', 'done': True, 'done_reason': 'stop', 'total_duration': 1777528400, 'load_duration': 239478000, 'prompt_eval_count': 163, 'prompt_eval_duration': 1055180700, 'eval_count': 22, 'eval_duration': 418785300, 'logprobs': None, 'model_name': 'llama3.2', 'model_provider': 'ollama'}, id='lc_run--019dbed9-7ad0-72e0-ba03-1dafe5a72af2-0', tool_calls=[{'name': 'multiply', 'args': {'a': 5, 'b': 25}, 'id': '749ef574-85fb-4ad9-8200-14a31d40ea37', 'type': 'tool_call'}], invalid_tool_calls=[], usage_metadata={'input_tokens': 163, 'output_tokens': 22, 'total_tokens': 185})

In [25]:
messages.append(result)

In [27]:
tool_result = multiply.invoke(result.tool_calls[0])
tool_result

ToolMessage(content='125', name='multiply', tool_call_id='749ef574-85fb-4ad9-8200-14a31d40ea37')

In [28]:
messages.append(tool_result)

In [30]:
print(llm.invoke(messages).content)

The result of multiplying 5 and 25 is 125.
